### ISSUE TITLE
Network can not learn features by using custom backbone & decode head

### ISSUE THREAD

**Issue by xDooooot** [open-mmlab/mmsegmentation#112]
MMsegmentation is a nice work and I've got a good result on DeepLab V3 with my custom dataset.
I encountered this error when I tried to implement UNet. The mIOU of validation set is always around 20% during the whole training procedure.
Here is my code of UNet. Thanks for any help.

```python
# unet_backbone.py
import torch
import torch.nn as nn
import torch.nn.functional as F
from mmcv.cnn import kaiming_init, build_conv_layer
from mmcv.cnn import ConvModule

from ..builder import BACKBONES


class PAM_Module(nn.Module):
    """ Position attention module"""
    #Ref from SAGAN
    def __init__(self, in_dim, conv_cfg=None, norm_cfg=None, act_cfg=None):
        super(PAM_Module, self).__init__()
        self.chanel_in = in_dim

        # self.query_conv = nn.Conv2d(in_channels=in_dim , out_channels=in_dim//8, kernel_size=1)
        # self.key_conv = nn.Conv2d(in_channels=in_dim , out_channels=in_dim//8, kernel_size=1)
        # self.value_conv = nn.Conv2d(in_channels=in_dim , out_channels=in_dim, kernel_size=1)
        self.query_conv = ConvModule(in_channels=in_dim, out_channels=in_dim//8, kernel_size=1,
                                     conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.key_conv = ConvModule(in_channels=in_dim, out_channels=in_dim//8, kernel_size=1,
                                   conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.value_conv = ConvModule(in_channels=in_dim, out_channels=in_dim, kernel_size=1,
                                     conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.gamma = nn.Parameter(torch.zeros(1))

        self.softmax = nn.Softmax(dim=-1)

    def forward(self,x):
        m_batchsize, C, width, height = x.size()
        proj_query = self.query_conv(x).view(m_batchsize,-1,width*height).permute(0,2,1)
        proj_key = self.key_conv(x).view(m_batchsize,-1,width*height)
        energy = torch.bmm(proj_query,proj_key)
        attention = self.softmax(energy)
        proj_value = self.value_conv(x).view(m_batchsize,-1,width*height)

        out = torch.bmm(proj_value,attention.permute(0, 2, 1))
        out = out.view(m_batchsize,C,width,height)

        out = self.gamma*out + x
        return out


class CAM_Module(nn.Module):
    """ Channel attention module"""
    def __init__(self, in_dim):
        super(CAM_Module, self).__init__()
        self.chanel_in = in_dim

        self.gamma = nn.Parameter(torch.zeros(1))
        self.softmax  = nn.Softmax(dim=-1)

    def forward(self,x):
        m_batchsize, C, width, height = x.size()
        proj_query = x.view(m_batchsize,C,-1)
        proj_key = x.view(m_batchsize,C,-1).permute(0,2,1)
        energy = torch.bmm(proj_query,proj_key)
        energy_new = torch.max(energy,-1,keepdim=True)[0].expand_as(energy)-energy
        attention = self.softmax(energy_new)
        proj_value = x.view(m_batchsize,C,-1)

        out = torch.bmm(attention,proj_value)
        out = out.view(m_batchsize,C,width,height)

        out = self.gamma*out + x
        return out

class DANetHead(nn.Module):
    def __init__(self, in_channels, out_channels, conv_cfg=None,
                 norm_cfg=dict(type='BN', requires_grad=True), act_cfg=dict(type='ReLU')):
        super(DANetHead, self).__init__()
        inter_channels = in_channels // 4
        self.conv5a = ConvModule(in_channels=in_channels, out_channels=inter_channels, kernel_size=3,
                                 padding=1, bias=False, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)

        self.conv5c = ConvModule(in_channels=in_channels, out_channels=inter_channels, kernel_size=3,
                                 padding=1, bias=False, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)

        self.sa = PAM_Module(in_dim=inter_channels)
        self.sc = CAM_Module(in_dim=inter_channels)
        self.conv51 = ConvModule(in_channels=inter_channels, out_channels=inter_channels, kernel_size=3,
                                 padding=1, bias=False, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.conv52 = ConvModule(in_channels=inter_channels, out_channels=inter_channels, kernel_size=3,
                                 padding=1, bias=False, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)

        self.conv8 = nn.Sequential(nn.Dropout2d(0.1, False),
                                   ConvModule(in_channels=inter_channels, out_channels=out_channels, kernel_size=1,
                                              conv_cfg=None, norm_cfg=None, act_cfg=None))

    def forward(self, x):
        feat1 = self.conv5a(x)
        sa_feat = self.sa(feat1)
        sa_conv = self.conv51(sa_feat)

        feat2 = self.conv5c(x)
        sc_feat = self.sc(feat2)
        sc_conv = self.conv52(sc_feat)

        feat_sum = sa_conv + sc_conv

        x = self.conv8(feat_sum)

        return x


class AttentionEncoder(nn.Module):
    def __init__(self, in_channels, out_channels,
                 conv_cfg=None, norm_cfg=dict(type='BN', requires_grad=True), act_cfg=dict(type='ReLU')):
        super(AttentionEncoder, self).__init__()
        self.down_conv = nn.Sequential(
            ConvModule(in_channels=in_channels, out_channels=out_channels, kernel_size=3,
                       padding=1, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg),
            ConvModule(in_channels=out_channels, out_channels=out_channels, kernel_size=3,
                       padding=1, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg),
        )
        self.pool = nn.MaxPool2d(kernel_size=2, ceil_mode=True)

    def forward(self, x):
        x = self.down_conv(x)
        return self.pool(x)


class AttentionDecoder(nn.Module):
    def __init__(self, in_channels, mid_channels, out_channels,
                 conv_cfg=None, norm_cfg=dict(type='BN', requires_grad=True), act_cfg=dict(type='ReLU')):
        super(AttentionDecoder, self).__init__()
        self.conv = nn.Sequential(
            ConvModule(in_channels=in_channels, out_channels=mid_channels, kernel_size=3,
                       padding=1, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg),
            DANetHead(in_channels=mid_channels, out_channels=mid_channels, conv_cfg=conv_cfg,
                      norm_cfg=norm_cfg, act_cfg=act_cfg),
            nn.ConvTranspose2d(in_channels=mid_channels, out_channels=out_channels, kernel_size=4, stride=2, padding=1),
        )

    def forward(self, x_copy, x):
        x = torch.cat((x_copy, x), 1)
        return self.conv(x)


@BACKBONES.register_module()
class AttentionBaseUNet(nn.Module):
    def __init__(self, in_channels, conv_cfg=None,
                 norm_cfg=dict(type='BN', requires_grad=True), act_cfg=dict(type='ReLU')):
        super(AttentionBaseUNet, self).__init__()
        self.down1 = AttentionEncoder(in_channels, 64, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.down2 = AttentionEncoder(64, 128, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.down3 = AttentionEncoder(128, 256, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.down4 = AttentionEncoder(256, 512, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.mid_conv = nn.Sequential(
            ConvModule(512, 1024, kernel_size=3, padding=1, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg),
            ConvModule(1024, 1024, kernel_size=3, padding=1, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg),
        )
        self.up1 = AttentionDecoder(1536, 512, 512, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.up2 = AttentionDecoder(768, 256, 256, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.up3 = AttentionDecoder(384, 128, 128, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)
        self.up4 = AttentionDecoder(192, 64, 64, conv_cfg=conv_cfg, norm_cfg=norm_cfg, act_cfg=act_cfg)

    def init_weights(self, pretrained=None):
        pass

    def forward(self, x):
        x1 = self.down1(x)
        x2 = self.down2(x1)
        x3 = self.down3(x2)
        x4 = self.down4(x3)
        x = self.mid_conv(x4)
        x = self.up1(x4, x)
        x = self.up2(x3, x)
        x = self.up3(x2, x)
        x = self.up4(x1, x)
        return x
```

```python
# unet_head.py
from ..builder import HEADS
from .decode_head import BaseDecodeHead
from ..backbones.unet_backbone import AttentionDecoder

@HEADS.register_module()
class Unet_head(BaseDecodeHead):
    def __init__(self, **kwargs):
        super(Unet_head, self).__init__(**kwargs)

    def forward(self, x):
        return self.cls_seg(x)
```

**Comment by xvjiarui** [comment]
Hi @xDooooot 
I am not sure what goes wrong. But you may check the following. 
1. The loss and acc during training. You may adjust the learning rate accordingly. 
2. You may check the input resolution of `Unet_head`, compare it with your input size.

**Comment by xDooooot** [comment]
> Hi @xDooooot
> I am not sure what goes wrong. But you may check the following.
> 
> 1. The loss and acc during training. You may adjust the learning rate accordingly.
> 2. You may check the input resolution of `Unet_head`, compare it with your input size.

Thanks, I'll try it. :)

**Comment by xvjiarui** [comment]
Hi @xDooooot 
Any updates? Have you figured it out?

**Comment by xDooooot** [comment]
Hi @xvjiarui 
**Sorry for updating so late.**

I've tried it out following your advice, but still stuck at the same problem.
Firstly I checked the input resolution of `Unet_head` and found no error in this way. It was at the same size as the input size. Then I reduced the learning rate from 1e-2 to 1e-3, 1e-4, 5e-5, and 1e-5, the problem remained.

And I found something interesting. If I removed the `DANetHead` module from the model, the mIOU of the validation set will go up during the training procedure. That's wired.

New codes as bellow.
```python
# Unet_backbone.py

import torch
import torch.nn as nn
import torch.nn.functional as F
from mmcv.cnn import kaiming_init, build_conv_layer
from mmcv.cnn import ConvModule

from ..builder import BACKBONES

class DoubleConvs(nn.Module):
    def __init__(self,
                 in_channels,
                 out_channels,
                 mid_channels=None,
                 conv_cfg=None,
                 norm_cfg=dict(type='BN', requires_grad=True)):
        super(DoubleConvs, self).__init__()
        self.conv_cfg = conv_cfg
        self.norm_cfg = norm_cfg
        if mid_channels is None:
            mid_channels = out_channels
        self.double_convs = nn.Sequential(
            ConvModule(
                in_channels=in_channels,
                out_channels=mid_channels,
                kernel_size=3,
                padding=1,
                conv_cfg=self.conv_cfg,
                norm_cfg=self.norm_cfg
            ),
            ConvModule(
                in_channels=mid_channels,
                out_channels=out_channels,
                kernel_size=3,
                padding=1,
                conv_cfg=self.conv_cfg,
                norm_cfg=self.norm_cfg
            )
        )

    def forward(self, x):
        return self.double_convs(x)

class Down_module(nn.Module):
    def __init__(self,
                 in_channels,
                 out_channels,
                 conv_cfg=None,
                 norm_cfg=dict(type='BN', requires_grad=True)):
        super(Down_module, self).__init__()
        self.conv_cfg = conv_cfg
        self.norm_cfg = norm_cfg

        self.down = DoubleConvs(in_channels, out_channels, None, self.conv_cfg, self.norm_cfg)
        self.pool = nn.MaxPool2d(kernel_size=2, ceil_mode=True)

    def forward(self, x):
        x = self.down(x)
        x_pooled = self.pool(x)
        return x, x_pooled

class Up_module(nn.Module):
    def __init__(self,
                 in_channels,
                 out_channels,
                 bilinear=False,
                 align_corners=True,
                 conv_cfg=None,
                 norm_cfg=dict(type='BN', requires_grad=True)):
        super(Up_module, self).__init__()
        self.conv_cfg = conv_cfg
        self.norm_cfg = norm_cfg
        self.bilinear = bilinear
        self.align_corners = align_corners

        if self.bilinear:
            # self.up = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=self.align_corners)
            # self.conv = DoubleConvs(in_channels, out_channels, in_channels//2, self.conv_cfg, self.norm_cfg)
            raise NotImplementedError
        else:
            self.up = nn.ConvTranspose2d(in_channels, out_channels, kernel_size=2, stride=2)
            self.up_conv = DoubleConvs(in_channels, out_channels, None, self.conv_cfg, self.norm_cfg)

    def forward(self, x_copy, x, interpolate=True):
        x = self.up(x)
        if interpolate:
            x = F.interpolate(x, size=(x_copy.size(2), x_copy.size(3)),
                              mode='bilinear', align_corners=True)
        else:
            diffY = x_copy.size(2) - x.size(2)
            diffX = x_copy.size(3) - x.size(3)
            x = F.pad(x, (diffX//2, diffX - diffX//2,
                          diffY//2, diffY-diffY//2))

        x = torch.cat([x_copy, x], dim=1)
        x = self.up_conv(x)
        return x

@BACKBONES.register_module()
class UnetBackbone(nn.Module):
    '''Implementation of original Unet head for encode and decode process.
    Must use with backbones/unet_backbone.py.
    '''
    def __init__(self,
                 in_channels,
                 bilinear=True,
                 align_corners=True,
                 conv_cfg=None,
                 norm_cfg=dict(type='BN', requires_grad=True)
                 ):
        super(UnetBackbone, self).__init__()
        self.in_channels = in_channels
        self.bilinear = bilinear
        self.conv_cfg = conv_cfg
        self.norm_cfg = norm_cfg
        self.align_corners = align_corners

        self.down1 = Down_module(in_channels, 64)
        self.down2 = Down_module(64, 128)
        self.down3 = Down_module(128, 256)
        self.down4 = Down_module(256, 512)

        self.middle_conv = DoubleConvs(512, 1024)

        self.up1 = Up_module(1024, 512)
        self.up2 = Up_module(512, 256)
        self.up3 = Up_module(256, 128)
        self.up4 = Up_module(128, 64)

        self.init_weights()

    def init_weights(self, pretrained=None):
        for m in self.modules():
            if isinstance(m, nn.ConvTranspose2d):
                kaiming_init(m)
            elif isinstance(m, nn.Conv2d) or isinstance(m, nn.Linear):
                kaiming_init(m)
                if m.bias is not None:
                    m.bias.data.zero_()
            elif isinstance(m, nn.BatchNorm2d):
                m.weight.data.fill_(1)
                m.bias.data.zero_()

    def forward(self, x):
        x1, x = self.down1(x)
        x2, x = self.down2(x)
        x3, x = self.down3(x)
        x4, x = self.down4(x)
        x = self.middle_conv(x)
        x = self.up1(x4, x)
        x = self.up2(x3, x)
        x = self.up3(x2, x)
        x = self.up4(x1, x)

        return x
```

Next, I will check this out and hope I can fix it.

**Again, thanks for your help. It was very useful.** 👍
